core/src/Cuda/Kokkos_Cuda_Instance.hpp

```diff
@@ -145,11 +148,11 @@ class CudaInternal {
   bool was_initialized = false;
   bool was_finalized   = false;
 
-  // FIXME_CUDA: these want to be per-device, not per-stream...  use of 'static'
-  //  here will break once there are multiple devices though
-  inline static unsigned long* constantMemHostStaging = nullptr;
-  inline static cudaEvent_t constantMemReusable       = nullptr;
-  inline static std::mutex constantMemMutex;
+  inline static std::set<int> cuda_devices = {};
+  inline static std::map<int, unsigned long*> constantMemHostStagingPerDevice =
+      {};
+  inline static std::map<int, cudaEvent_t> constantMemReusablePerDevice = {};
+  inline static std::map<int, std::mutex> constantMemMutexPerDevice;
```

Comment:
It's convenient to store the `cuda_devices` in a static variable so we can iterate over them. `set` might not be the most performant choice here but is convenient and it shouldn't matter much since we don't use it in performance-critical code. Similarly, `std::map` is mostly convenient.